src/domain/stock-market.ts

```diff
@@ -0,0 +1,59 @@
+import { logger } from '../config/logger';
+import { Trader } from './trader';
+import { isTickerSymbolValid, fetchStockPriceAt, isTradeDateValid } from '@server/domain/common';
+
+// Stock Market Class handling all event sourcing
+
+class StockMarket {
+    traderMap: Map<[trader_id: string], Trader> = new Map()
+
+    createTrader(name: string) {
+        if (name.length > 10) {
+            throw new Error("Trader name cannot be longer than 100 characters");
+        }
+
+        const trader = new Trader(name);
+        this.traderMap[trader.id] = trader;
+
+        logger.info(`Successfully created new trader ${name}`)
+
+        return { trader_id: trader.id };
+    }
+
+    async buy(trader_id: string, symbol: string, date: Date, quantity: number) {
+        if (!isTickerSymbolValid(symbol)) {
+            throw new Error(`Invalid ticker symbol ${symbol}`);
+        }
+
+        if (!isTradeDateValid(trader_id, date)) {
+            throw new Error('Date needs to be on or after the previous buy/sell date');
+        }
+
+        const price: any = await fetchStockPriceAt(symbol, date);
```

Comment:
Possible approach here would be to inject the stock price service into this StockMarket service.

EG `async buy(requestData: { trader_id: string, symbol: string, date: Date, quantity: number}, stockPriceService: IStockPriceService )`

Would be helpful for mocking for tests. And having an interface defined would allow for swapping easily to a different stockPriceService (eg change vendors in the future).